### PR TITLE
γ (F3) — Remove custom-metrics framework from recipes (delete src/utils/finetune_custom_metrics.py)

### LINKED ISSUE #526: γ (F3) — Remove custom-metrics framework from recipes (delete src/utils/finetune_custom_metrics.py)
Part of #465 (wrapper-only principle). Carved out of closed PR #523.

## Context

The custom recipes patch `__init__()` to read `cfg.calculate_custom_metrics` and `cfg.custom_metrics`, then patch the loss-compute path to call user-supplied metric functions from `src/utils/finetune_custom_metrics.py`. The #465 audit identified this as a high-risk class: a buggy custom metric could (and did) crash a training run mid-flight, and the patch sits squarely inside `train()`'s loop body.

The framework currently has **no live consumers** in this repo (verified during the audit). Per the wrapper-only policy, anything that needs to read/write training-loop state belongs in a wrapper, not in `train()`. Since nothing uses the recipe-side hook, the simplest path is to delete the framework entirely; a wrapper-layer reimplementation can come back later if a real consumer appears.

## Scope

- **Recipes affected**: `lora_finetune_single_device_stable`, `lora_finetune_single_device_nightly`, `lora_finetune_distributed_stable`
  - Drop module-level \`CUSTOM_METRICS_AVAILABLE\` block
  - Drop \`__init__\` config-reads + \`self._calculate_custom_metrics\` / \`self._custom_metrics\`
  - Drop the per-step compute call in \`compute_loss()\` / \`_loss_step()\`
- **File deleted**: \`src/utils/finetune_custom_metrics.py\` (~73 lines)
- **YAML refs affected**: drop \`calculate_custom_metrics\` / \`custom_metrics\` from \`finetune_template.yaml\` if present
- **\`setup_finetune.py\`**: drop argparse / merge entries for these flags if present
- **Tests affected**: any custom-metrics unit tests (verify nothing else imports the deleted module)
- **User-facing config flags removed**: \`calculate_custom_metrics\`, \`custom_metrics\`

## Reference implementation

[PR #523](https://github.com/niznik-dev/cruijff_kit/pull/523) (closed) contains the working diff. The branch `465-strict-wrapper-principle` can be cherry-picked from.

## Acceptance criteria

- [ ] \`grep -n \"cruijff_kit patch\" src/tools/torchtune/custom_recipes/*.py\` returns no F3-related markers
- [ ] \`src/utils/finetune_custom_metrics.py\` is gone; \`grep -rn finetune_custom_metrics src/ tests/\` returns nothing
- [ ] \`pytest tests/unit/\` passes
- [ ] CHANGELOG note flagging the removal (no consumer impact at rip time, but it is a public surface)

—MxC

### PR DESCRIPTION
**Pull request by niznik-dev**
Closes #526

# Description

Third rip of milestone [Wrapper-only principle (#465)](https://github.com/niznik-dev/cruijff_kit/milestone/11) — deletes the custom-metrics framework in full. The #465 audit flagged this as the highest-risk in-\`train()\` patch (a buggy metric can corrupt loss accounting or kill a run mid-step). With **no live consumers** in this repo and a defensive try/except already in place, the simplest path is full deletion; a wrapper-layer reimplementation can come back later if needed.

**Stacked on #531** (β / F1). PR base is \`525-wrapper-layer-f1\` so the diff shows only F3 changes. Will retarget to \`main\` after #531 (and #530) merge.

## What was removed

| Surface | Detail |
|---|---|
| Recipe patches | 12 patch blocks across the 3 recipes (import + \`__init__\` + \`_loss_step\` / step body + log emission). Nightly's log block was missing its \`# !--- cruijff_kit patch ---!\` markers but was functionally F3; removed alongside the marked ones. |
| Module | \`src/utils/finetune_custom_metrics.py\` (~74 lines) — 0 importers outside the three recipes |
| \`docs/ARCHITECTURE.md\` | Removed 4 references: directory tree entry, custom-features bullet, utilities catalogue, and the obsolete "Extension Points / Adding Custom Metrics" section |
| \`CLAUDE.md\` | **Second rewrite of the #465 policy block.** With F3 gone, custom-metrics was the last live in-\`train()\` example. Block now states the principle abstractly ("loss accounting, gradient state, checkpoint scheduling, per-step metric emission") and tightens the forward-pointer to the milestone. Tone shifts from "defensive guards required" to "not allowed" — accurate after this rip, since silent-corruption modes survive even with guards. |
| \`CHANGELOG.md\` | New \`[Unreleased] / Removed\` entry (#465, #526) |

## What's intentionally untouched

- **YAML** — custom-metrics was never exposed as a config knob
- **\`setup_finetune.py\`** — never a CLI flag, never in RECIPE_PARAM_MAPPING
- **Tests** — **no F3 test coverage existed.** Worth noting on its own: a feature that touches training-loop math shipped with zero unit tests, which is part of why the #465 audit caught it as high-risk

## New Dependencies

None.

# Testing Instructions

- \`grep -rn "calculate_custom_metrics\\|finetune_custom_metrics\\|CUSTOM_METRICS_AVAILABLE\\|_custom_metrics\\|_calculate_custom_metrics" src/ tests/ docs/ blueprints/ .claude/ CLAUDE.md\` — empty
- \`grep -rn "calculate_custom_metrics\\|finetune_custom_metrics" CHANGELOG.md\` — only the new Unreleased entry + historical 0.3.2 line
- \`grep -c "cruijff_kit patch" src/tools/torchtune/custom_recipes/lora_finetune_*.py\` — stable / distributed drop by 8 marker lines each; nightly drops by 6 (the 4th block was unmarked)
- \`ruff check\` + \`ruff format --check\` on all touched files — clean
- \`python -m py_compile\` clean on all three recipes
- \`pytest tests/unit/\` — 1157 passed (same as F1; no F3 tests existed to delete)
- **SLURM smoke (deferred to combined F2+F1+F3 multi-epoch 1B run after #530 merges)**: confirms recipes train cleanly without the custom-metrics path, every \`epoch_N/\` dir is produced, and \`post_finetune.py\` rewrites \`adapter_config.json\`. Covers Marin's 💭 from PR #531.

—MxC

### CODE REVIEW

File: src/tools/torchtune/custom_recipes/lora_finetune_single_device_nightly.py [hunk 1]
```diff
@@ -882,11 +843,6 @@ def train(self) -> None:
                             if self._clip_grad_norm is not None:
                                 log_dict.update({"grad_norm": grad_norm})
```
Comment (niznik-dev) [inline src/tools/torchtune/custom_recipes/lora_finetune_single_device_nightly.py:844]: 🧭 [**Marin**](https://github.com/niznik-dev/llm-personas/blob/main/personas/marin/marin.md) 🧭

📍 **look here.** Just above this line, nightly used to have a small block that looked like:

```python
# Add custom metrics to log_dict and then reset for the next step
if self._custom_metrics:
    log_dict.update(self._custom_metrics)
    self._custom_metrics = {}
```

…**without** the `# !--- cruijff_kit patch ---!` guards that wrap the other three F3 blocks in this file. It got removed alongside the marked blocks because it was functionally F3 code, but it's the one block in the rip that wouldn't have shown up in a `grep cruijff_kit patch` audit.

No concern for this PR — the deletion is clean — but it's the kind of thing the `# !--- cruijff_kit patch ---!` convention exists to prevent. Worth a one-line follow-up note: any unmarked nightly-only divergence still needs the markers, even when the corresponding stable block has them.

File: CLAUDE.md [hunk 2]
```diff
@@ -62,11 +62,9 @@ For detailed architecture documentation, see [ARCHITECTURE.md](docs/ARCHITECTURE
 All cruijff_kit divergences from torchtune in `src/tools/torchtune/custom_recipes/` live in marked `# !--- cruijff_kit patch ---!` blocks — those blocks are the audit trail. Risk is governed by **location**, not patch count:
 
 - **Outside `train()`** (imports, config reads, `_setup_data`, post-`save_checkpoint` file ops): fine where they are. Failure modes are detectable — missing config keys raise at init, file-ops fail loudly.
-- **Inside `train()` or hooks that read/write training-loop state** (e.g., `_loss_step`, custom metrics calculation): require **defensive guards** — init-time validation that fails loud on bad config, runtime assertions on invariants, and a dedicated unit test covering the patched path.
+- **Inside `train()` or hooks that read or write training-loop state** (loss accounting, gradient state, checkpoint scheduling, per-step metric emission): not allowed. Silent-corruption modes — zero-checkpoint runs, mid-step crashes that taint the metric stream — only appear here, and they survive even with defensive guards.
```
Comment (niznik-dev) [inline CLAUDE.md]: 🧭 [**Marin**](https://github.com/niznik-dev/llm-personas/blob/main/personas/marin/marin.md) 🧭

📍 **look here.** Two intentional shifts in this paragraph since PR #531's rewrite:

1. **Tone**: was "require defensive guards" (you may patch if you guard carefully). Now "not allowed" (don't patch in-`train()` at all). This is the actual delta the milestone is producing — and it matches the lesson F3 surfaced, which is that even a guarded patch (custom-metrics had a try/except auto-disable) can corrupt the metric stream silently before the guard catches the failure.

2. **Examples**: switched from naming live patches (`_loss_step`, custom metrics) to naming **categories** of training-loop state (loss accounting, gradient state, checkpoint scheduling, per-step metric emission). Right call now that the milestone is making the live examples disappear — generic categories don't go stale rip by rip.

Forward-pointer paragraph below stays accurate while F4 + F5 + F6 are still in flight. After milestone close, this whole block could collapse to a one-liner or be retired entirely; flagging that as a milestone-close follow-up rather than something this PR should do.

**Review by niznik-dev — COMMENTED** [review]
🧭 [**Marin**](https://github.com/niznik-dev/llm-personas/blob/main/personas/marin/marin.md) 🧭

Third stop on the [Wrapper-only principle (#465)](https://github.com/niznik-dev/cruijff_kit/milestone/11) tour. The **γ rip** — custom-metrics framework gone in full. Even tighter than #531: +3 / −226 across seven files. Nothing to add, almost everything to delete.

📍 **Stack state.** PR is stacked **three deep**: `main` → `524-wrapper-layer-f2` (PR #530, α / F2) → `525-wrapper-layer-f1` (PR #531, β / F1) → `526-wrapper-layer-f3` (this PR). Base = `525-wrapper-layer-f1`. Each PR retargets to `main` one click at a time as the one below it merges. Worth tracking when CI re-runs against new bases.

**Reading order I'd suggest:**

1. One recipe (`lora_finetune_single_device_stable.py`) — four marked F3 blocks gone (import, `__init__`, `_loss_step`, log emission). This is the pattern.
2. `lora_finetune_distributed_stable.py` — same pattern, same shape.
3. `lora_finetune_single_device_nightly.py` — same pattern, **with one wrinkle**. Inline 📍 below.
4. `src/utils/finetune_custom_metrics.py` — gone. Confirm no surviving importer.
5. `CLAUDE.md` — second rewrite of the #465 policy block. Inline 📍 — the tone has shifted.
6. `docs/ARCHITECTURE.md` + `CHANGELOG.md` — small follow-on.

---

### Per-file overview

| File | What this PR does to it | Skim/Read/Pause |
|---|---|---|
| `src/tools/torchtune/custom_recipes/lora_finetune_single_device_stable.py` | The canonical F3 surgery: four marked patch blocks gone. After this PR the recipe's `_loss_step` is identical to upstream torchtune. | **read** |
| `src/tools/torchtune/custom_recipes/lora_finetune_distributed_stable.py` | Same four-block pattern as stable, with the `_loss_step` equivalent inline in the step body (distributed has no separate method). | skim |
| `src/tools/torchtune/custom_recipes/lora_finetune_single_device_nightly.py` | Same three marked patch blocks as stable; **fourth (log-emission) block was unmarked** but functionally F3. Removed alongside the marked ones. Inline 📍. | **read** |
| `src/utils/finetune_custom_metrics.py` | Whole file deleted (~74 lines). Confirmed zero importers outside the three recipes. | skim |
| `docs/ARCHITECTURE.md` | Removed 4 references: directory tree entry, custom-features bullet, utilities catalogue, and the entire "Extension Points / Adding Custom Metrics" section. | skim |
| `CLAUDE.md` | **Second rewrite of the #465 policy block** — the in-`train()` example list lost its last live citation. Block now phrased abstractly + tone shifted from "defensive guards required" to "not allowed." Inline 📍. | **pause** |
| `CHANGELOG.md` | One new `[Unreleased] / Removed` entry. Historical 0.3.2 line intentionally untouched (it documents an earlier defensive hardening of the now-deleted code). | skim |

---

### Severity callouts

🟡 **watch — zero pre-existing tests for F3.** The framework shipped with **no unit-test coverage**. Mattie called this out in the PR body, but it's worth weighting: a feature that touched training-loop math passed code review at some prior point with no tests. That history won't repeat under the wrapper-only policy (which forbids new in-`train()` patches), but it informs how to think about the absence — there's no regression set to break here, but there was also never one to catch a regression.

🟡 **watch — base branch is `525-wrapper-layer-f1`.** Same as #531: retarget to `main` once the upstream PRs (#530, then #531) merge. CI re-runs against the new base each retarget.

🟢 **note — the recipes' loss step is now identical to upstream torchtune.** After this PR + #531, the only remaining in-`train()` divergence is the tqdm `miniters` kwarg (F5 / #527). Loss math, gradient accounting, and checkpoint scheduling all match upstream verbatim. That's a meaningful safety property — anyone reading the recipe doesn't need to ask "is this our patched math or theirs?" anymore.

💭 **question — wrapper-layer reimplementation of custom metrics?** PR body says "a wrapper-layer reimplementation can come back later if needed." Worth asking now while the decision is fresh: if a real metrics consumer appears, the wrapper-side hook needs to either (a) re-run forward pass on the post-train checkpoint (cheap, no training-loop intrusion) or (b) accumulate metrics that the recipe emits to disk during training (would need a small log emission patch, but a write-only one, not a math-touching one). Probably worth a follow-up issue describing the constraint, even if no implementation is planned.

---

Mattie did the rip; I'm just walking the path. Trust your own read.

**Comment by msalganik** [comment]
We discussed this. Looks good to me, pending minor changes we discussed. I don't need to see it again before merge. But happy to talk if it would be helpful.
